Upgrade waku specs to waku/1

cammellos opened the pull request:
This commit upgrades the specs to `waku/1` and marks `waku/0` as
deprecated.

Other changes include:
    
- RLP code are now 0,1,2,3,4,5 for status-options
- Version is removed from handshake
- Waku/1 is now advertised

`waku-1.md` is seen as a new file, so it's a bit difficult to see the diff, github does not show it (`gitk` does though, in case you use it).

Pasting here the output of `diff`
```
3,4c3,4
< version: 0.6.0
< status: Deprecated
---
> version: 1.0.0
> status: Draft
46c46
<     - [Differences between shh/6 waku/0](#differences-between-shh6-waku0)
---
>     - [Differences between shh/6 waku/1](#differences-between-shh6-waku1)
72a73,74
> This protocol needs to advertise the `waku/1` [capability](https://ethereum.gitbooks.io/frontier-guide/devp2p.html).
> 
100,105c102,107
< pow-requirement-key = 48
< bloom-filter-key = 49
< light-node-key = 50
< confirmations-enabled-key = 51
< rate-limits-key = 52
< topic-interest-key = 53
---
> pow-requirement-key = 0
> bloom-filter-key = 1
> light-node-key = 2
> confirmations-enabled-key = 3
> rate-limits-key = 4
> topic-interest-key = 5
116c118
< status = "[" version status-options "]"
---
> status = status-options
120,122d121
< ; version is "an integer (as specified in RLP)"
< version = DIGIT
< 
396a396,399
> 
> The currently advertised capability is `waku/1`. This needs to be advertised in the `hello` `ÐΞVp2p` [packet](https://ethereum.gitbooks.io/frontier-guide/devp2p.html).
> If a node supports multiple versions of `waku`, those needs to be explicitly advertised. For example if both `waku/0` and `waku/1` are supported, both `waku/0` and `waku/1` MUST to be advertised.
> 
409c412
< - We enable bridging between `shh/6` and `waku/0` until such a time as when we are ready to gracefully drop support for `shh/6` (1, 2, 3).
---
> - We enable bridging between `shh/6` and `waku/1` until such a time as when we are ready to gracefully drop support for `shh/6` (1, 2, 3).
420c423
< `waku/0` and `shh/6` are different DevP2P subprotocols, however they share the same data format making their envelopes compatible. This means we can bridge the protocols naively, this works as follows.
---
> `waku/1` and `shh/6` are different DevP2P subprotocols, however they share the same data format making their envelopes compatible. This means we can bridge the protocols naively, this works as follows.
436c439
< It is desirable to have a strategy for maintaining forward compatibility between `waku/0` and future version of waku. Here we outline some concerns and strategy for this.
---
> It is desirable to have a strategy for maintaining forward compatibility between `waku/1` and future version of waku. Here we outline some concerns and strategy for this.
519c522,524
< ### Version 0.6
---
> ### Version 1
> 
> Released []()
521c526,528
< - Mark spec as Deprecated mode in terms of its lifecycle.
---
> - Removed `version` from handshake
> - Changed `RLP` keys from 48,49.. to 0,1..
> - Upgraded to `waku/1`
573c580
< ### Differences between shh/6 and waku/0
---
> ### Differences between shh/6 and waku/1
577c584
< - RLPx subprotocol is changed from `shh/6` to `waku/0`.
---
> - RLPx subprotocol is changed from `shh/6` to `waku/1`.

```

kdeme commented on specs/waku/waku.md:
```diff
@@ -519,6 +520,20 @@ Known static nodes MAY also be used.
 
 - Clarify the preferred way of handling unknown keys in the `status-options` association list.
```
This was put in changelog for when version 0.5 gets released.

kdeme commented on specs/waku/waku.md:
```diff
@@ -519,6 +520,20 @@ Known static nodes MAY also be used.
+Released []()
+
+- Marked as [deprecated](./waku-0.md)
+- Corrected RLP keys to their int values
```
This is not a specification change but mistake in implementation. Thus it should not be mentioned here imo.

cammellos replied:
Although I agree with you that it should not be there and that's correct, pragmatically we had to make a change to the specs (RLP `0` became `48`) a posteriori.

Essentially the way I see it is that we decided that the specs in this case should reflect the live implementation (which is not a "correct" choice, but it's a pragmatic one as we only have one live implementation), so I feel it's useful to justify the change.

Having said that I don't feel to strongly about it, so happy to remove it if you don't buy the argument :)

oskarth replied:
How about:

"Correct spec/implementation mismatch: Change RLP keys to be the their int values in order to reflect production behaviour"?

kdeme replied:
Oh, I hadn't even noticed that the integers were changed in the ABNF specification. Which shows that this flow of doing the duplication of the file and making changes in it in one go (in one PR) is not good. I'm basically forced to do a diff between files locally on my machine to spot all changes.

That being said, I don't agree with changing the values. We can work around the problem just in the status-go implementation (accepting both integers), no? And this way remove it later again. Instead of uglifying the spec for the rest of its lifetime.

kdeme replied:
OK, nevermind on changing the values. I see they were changed back in latest spec. Damn, confusing.
Can we please not do things like this because the implementation did something wrong?

kdeme replied:
Anyway, we can leave it like this if it must. No other client will claim to have been supporting version 0.5.
Just, it is not necessary to be there and it feels like a weird thing to do. I wouldn't want this to become a habit in the future.

cammellos replied:
> Which shows that this flow of doing the duplication of the file and making changes in it in one go (in one PR) is not good. I'm basically forced to do a diff between files locally on my machine to spot all changes.

That's fair, two PRs would have been a better way, I will split them.

> That being said, I don't agree with changing the values. We can work around the problem just in the status-go implementation (accepting both integers), no?

We can, it's more work, that means:
1) Make the handshake compatible between current botched version and actual v0 specs (which we will never use btw, as nimbus is not ready for prime time and status-go will jump to waku/1 directly likely)

2) Push out V1 removing the unused `version` field

As opposed to just 2) , specs might look not as "good", but is that really a valid point? developer working hours + complexity of code vs specs not looking aesthetically pleasing?

> Can we please not do things like this because the implementation did something wrong?

Not sure, if you have a better way to handle it I am all for it, as I said, this is clearly not the correct way, but pragmatically it's the one that will lend itself to the least amount of complexity  in the currently only live implementation and saves some dev hours.

Effectively whoever will implement `waku/0` now will be able to talk to the only currently live implementation, and once we push out `waku/1` similarly it should, which I think it's what matters most.

kdeme replied:
I still don't really understand why the spec needs the inbetween 0.5 update. You can just have a version 0 in status-go that is not compliant to the waku version 0 specification. At least, if then you are going to update it to version 1. I don't see the practical difference and thus don't understand why you say it is pragmatically the way to handle it, except for being correct according to the letter of the spec, that is by altering the spec.

But, as version 0.5 will not really be used and is just there for backwards compatibility in status-go, it is no big deal for me. No value in wasting more man hours discussing it :-).

Main issue here was the confusion created by doing the two version releases in one PR.

cammellos replied:
I agree that there's not value in discussing this any further, either ways I don't think it makes a difference, I am ok with leaving it as it or changing it, so I will leave it as it unless someone has a strong opinion.

kdeme commented on specs/waku/waku.md:
```diff
@@ -391,6 +389,9 @@ In later versions this will be amended by nodes communication threshholds, settl
 
 ## Upgradability and Compatibility
 
+The currently advertised capabilities is `waku/1`. This needs to be advertised in the `hello` `ÐΞVp2p` [packet](https://ethereum.gitbooks.io/frontier-guide/devp2p.html).
```
```suggestion
The currently advertised capability is `waku/1`. This needs to be advertised in the `hello` `ÐΞVp2p` [packet](https://ethereum.gitbooks.io/frontier-guide/devp2p.html).
```

kdeme commented on specs/waku/waku-0.md:
```diff
@@ -0,0 +1,600 @@
+Released []()
+
+- Marked as deprecated
+- Corrected RLP keys to their int values
```
Same as remark in other file

kdeme commented on specs/waku/waku-0.md:
```diff
@@ -0,0 +1,600 @@
+
+### Upcoming Version
+
+- Clarify the preferred way of handling unknown keys in the `status-options` association list.
```
Same as remark in other file

kdeme reviewed (changes requested):
I think if we want to do the flow of releasing like this, then perhaps we should first make the functional changes in e.g in this case a 0.5 version first, and then create the new file and bump it to 1.0  it in a separate PR? Not sure, but it would be easier to track the changes then.

kdeme commented on specs/waku/waku-0.md:
```diff
@@ -0,0 +1,600 @@
+
+rate-limits = "[" limit-ip limit-peerid limit-topic "]"
+
+pow-requirement-key = 48
```
Why must these be changed? Can't it be worked around in status-go temporarily?

kdeme replied:
Ah woops, this is the old specification. I was confused with comment lower down.

Still, old specification had the correct values. Why mess this up now? Just because status-go was not correct?

kdeme commented:
The procedure of when a new version is released should be very clear, else we might end up doing another change to the specification soon. So I decided to write down the way I understand it now.

Currently, we increase the version of the specification with a minor number or with a major number.

Devp2p capabilities allow only one integer for version, so we have only access to only a "major version" in the actual implementations.
This means, that we would only bump the major version if we have a breaking change. And then an implementation can decide to list which major versions it still supports. This is somewhat in line with how libraries do for their API. Major -> breaking change.

On any smaller changes, that are not breaking, the major version should not be changed (unless we want to force incompatibility for some reason?). In case of such update, implementations will not be able to decide if they support a peer without the update it as they simply won't know the specific minor version of that peer (except perhaps through watching the behaviour of its peer).

Now some examples of hypothetical changes to illustrate this:
1. Introduction of a new packet type to do a slightly different mail server request: Not a breaking change as unknown packet types should be ignored. Version is changed to 1.1. If a node does not want to remain connected with a peer that doesn't support this packet type, it will not be able to decide this at handshake. It might notice this when it gets no reply and decide to disconnect.
2. Some new option in the `status-options` passed at handshake. Older versions should ignore this unknown field, thus version can be changed to 1.2. Newer version would not see this new option when in contact with peers that don't support it. These newer peers have to use to the current way of working / current defaults. E.g. we want to add a larger bloom filter (yeah, not the best example). New k:v pair is added to the options list. Older client will only send the current bloom filter. Peers use still that one to select which envelopes are relayed.
3. After the introduction of the new packet type in version 1.1, it is decided to remove the old packet type for a mail server request. I think this is a breaking change as else clients with version 1.0 would have no way to do a mail server request. Thus the version is bumped to 2.0.
4. Some crucial bug is fixed. Some crucial implementation change, like PoW mechanism. Etc. All seems like bigger changes that would change to version 3.0.

Now:
1. Is my understanding correct? Or are there better/other examples that are not covered? Specifically where we need access to the minor version?
2. Is the specification currently clear enough to reflect this understanding?

cammellos commented:
> Is my understanding correct? Or are there better/other examples that are not covered? Specifically where we need access to the minor version?
I have the same understanding.

Currently the minor version is never communicated, I am ok with that for the time being as it can be assumed to be `0` or that it does not matter.
I prefer a non committal approach, so the further down the line we can defer the decision, the more likely we will get it right and we will have an actual problem that might be solved by specifying the minor version.

> Is the specification currently clear enough to reflect this understanding?

This I am not sure, I think we leave a bit room for interpretation, I am for giving clear examples, like the ones you gave, even though they are not "spec-like" maybe, as what matters most is that people reading this will understand. So I would add a section basically with something on the lines of what you wrote.

Samyoul commented:
If we are deprecating `waku/0` I'd suggest that we implement in `waku/1` that option keys are either always encoded as `uint` or we use strict type encoding [such as found in the RLP specs](https://github.com/ethereum/wiki/wiki/RLP#examples).

Do we still implement the changes to `status-go`'s `waku/0` handshake handling introduced in [my on going PR](https://github.com/status-im/status-go/pull/1931)?

cammellos commented:
@Samyoul 
> If we are deprecating waku/0 I'd suggest that we implement in waku/1 that option keys are either always encoded as uint

I believe that was already the case, both `waku/0` and `waku/1` mandated that keys ought to be interpreted as uint.

As I understand RLP encoding was correctly followed, but `status-go` was passing the wrong bytes over the wire (`0x48` instead of `0x00`, because it was mistakenly passing a string instead of an uint locally). 

What the spec is proposing is not to change `waku/0` current behavior, but rather address that in `waku/1` instead, this is some code that is both compatible with `waku/0` and `waku/1` https://github.com/status-im/status-go/tree/feature/upgrade-to-waku-1

Samyoul commented:
@cammellos 
>What the spec is proposing is not to change waku/0 current behavior, but rather address that in waku/1 instead, this is some code that is both compatible with waku/0 and waku/1 https://github.com/status-im/status-go/tree/feature/upgrade-to-waku-1

Ok so the [work I've done](https://github.com/status-im/status-go/pull/1931) to get the `status-go` to handshake with the nim implementation would be the beginning of `waku/1`? Or are we reimplementing `waku/0` but use uint keys rather than strings?

I suppose using versioning as suggested in [your branch](https://github.com/status-im/status-go/tree/feature/upgrade-to-waku-1) would make my work redundant. If you open the RP on that branch I can make some suggestions as to how we implement waku versioning functionality. Just things like abstraction of the version struct and factory pattern implementation for selecting the version.

cammellos commented:
> Or are we reimplementing waku/0 but use uint keys rather than strings?
I would say to leave `waku/0` as it is so we don't mess with existing clients etc, and just implement the fix with the spec changes in `waku/1`.

>  If you open the RP on that branch I can make some suggestions as to how we implement waku versioning
I will open the PR (albeit in draft mode as I haven't finished adding tests etc), that would be very much appreciated, thanks!

oskarth commented on specs/waku/waku.md:
```diff
@@ -42,7 +42,7 @@ redirect_from:
     - [Version 0.2](#version-02)
     - [Version 0.1](#version-01)
-    - [Differences between shh/6 waku/0](#differences-between-shh6-waku0)
+    - [Differences between shh/6 waku/1](#differences-between-shh6-waku0)
```
Link seems off

oskarth commented on specs/waku/waku.md:
```diff
@@ -68,6 +68,7 @@ Waku was created to incrementally improve in areas that Whisper is lacking in, w
 ### Use of DevP2P
 
 For nodes to communicate, they MUST implement devp2p and run RLPx. They MUST have some way of connecting to other nodes. Node discovery is largely out of scope for this spec, but see the appendix for some suggestions on how to do this.
+This protocol needs to advertise the `waku/1` [capabilities](https://ethereum.gitbooks.io/frontier-guide/devp2p.html).
```
```suggestion
This protocol needs to advertise the `waku/1` [capability](https://ethereum.gitbooks.io/frontier-guide/devp2p.html).
```

oskarth commented on specs/waku/waku.md:
```diff
@@ -391,6 +389,9 @@ In later versions this will be amended by nodes communication threshholds, settl
 ## Upgradability and Compatibility
 
+The currently advertised capability is `waku/1`. This needs to be advertised in the `hello` `ÐΞVp2p` [packet](https://ethereum.gitbooks.io/frontier-guide/devp2p.html).
+If a node supports multiple versions of `waku`, those needs to be explicitly advertised. For example if both `waku/0` and `waku/1` are supported, both `waku/0` and `waku/1` needs to be advertised.
```
```suggestion
If a node supports multiple versions of `waku`, those needs to be explicitly advertised. For example if both `waku/0` and `waku/1` are supported, both `waku/0` and `waku/1` MUST to be advertised.
```

oskarth commented on specs/waku/waku.md:
```diff
@@ -519,6 +520,20 @@ Known static nodes MAY also be used.
+Released []()
+
+- Marked as [deprecated](./waku-0.md)
+- Corrected RLP keys to their int values
```
I commented on different formulation here but seems like it got lost

cammellos committed:
Upgrade waku specs to waku/1

This commit upgrades the spec to waku/1.

Changes include:

- RLP code are now 0,1,2,3,4,5 for status-options
- Version is removed from handshake
- Waku/1 is now advertised

Co-Authored-By: Kim De Mey <kim.demey@gmail.com>

kdeme commented on specs/waku/waku-1.md:
```diff
@@ -0,0 +1,599 @@
+
+
+The currently advertised capability is `waku/1`. This needs to be advertised in the `hello` `ÐΞVp2p` [packet](https://ethereum.gitbooks.io/frontier-guide/devp2p.html).
+If a node supports multiple versions of `waku`, those needs to be explicitly advertised. For example if both `waku/0` and `waku/1` are supported, both `waku/0` and `waku/1` MUST to be advertised.
```
```suggestion
If a node supports multiple versions of `waku`, those needs to be explicitly advertised. For example if both `waku/0` and `waku/1` are supported, both `waku/0` and `waku/1` MUST be advertised.
```

kdeme commented on specs/waku/waku-1.md:
```diff
@@ -0,0 +1,599 @@
+
+It is desirable to have a strategy for maintaining forward compatibility between `waku/1` and future version of waku. Here we outline some concerns and strategy for this.
+
+- **Connecting to nodes with multiple versions:** The way this SHOULD be accomplished in the future is by negotiating the versions of subprotocols, within the `hello` message nodes transmit their capabilities along with a version. As suggested in [EIP-8](https://eips.ethereum.org/EIPS/eip-8), if a node connects that has a higher version number for a specific capability, the node with a lower numb ...
```
The last part of this sentence is contradicting with how we deal with the version number now, and should be removed or reformulated.

kdeme commented on specs/waku/waku-1.md:
```diff
@@ -0,0 +1,599 @@
+    - [Recommendations for clients](#recommendations-for-clients)
+    - [Node discovery](#node-discovery)
+- [Changelog](#changelog)
+    - [Version 0.5](#version-05)
```
Update to add link for Version 1?

kdeme commented on specs/waku/waku-0.md:
```diff
@@ -0,0 +1,592 @@
+    - [Recommendations for clients](#recommendations-for-clients)
+    - [Node discovery](#node-discovery)
+- [Changelog](#changelog)
+    - [Version 0.5](#version-05)
```
Update to add link for version 0.6?

cammellos commented:
@kdeme thanks for the feedback, addressed!

oskarth reviewed (dismissed):
Looks great, thanks!

Used as an excuse to install gitk, surprisingly nice :)

cammellos committed:
Update refs to 0.6 and 1.0

cammellos commented:
@kdeme could you re-review please? Thanks!

Samyoul reviewed (approved):
I don't have any objections to these changes. They line up with what we've been discussing.
